kernel/src/sched.rs

```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
-const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+pub(crate) const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+
+/// Trait which any scheduler must implement.
+pub trait Scheduler<C: Chip> {
+    /// Return the next process to run, and the timeslice length for that process.
+    /// Passing None instead of an AppId will lead to nothing being run.
+    /// Passing None instead of a timeslice will cause the process to be run cooperatively
+    fn next(&self) -> (Option<AppId>, Option<u32>);
+
+    /// Informs the scheduler of why the last process stopped executing, and how
+    /// long it executed for. Notably, execution_time_us will be `None`
+    /// if the the scheduler requested this process be run cooperatively.
+    fn result(&self, result: StoppedExecutingReason, execution_time_us: Option<u32>);
+
+    /// Tell the scheduler to execute kernel work such as interrupt bottom halves
+    /// and dynamic deferred calls. Most schedulers will use this default
+    /// implementation, but schedulers which at times wish to defer interrupt
+    /// handling will reimplement it.
+    ///
+    /// The reason for stubbing this out to allow schedulers to reimplement it
+    /// is that I (Hudson Ayers) believe that deferring bottom half
+    /// interrupt handling, or only servicing certain interrupts,
+    /// can be useful in helping userspace apps to meet
+    /// deadlines if a scheduler is given sufficient information about app
+    /// workloads. This stub could also potentially be useful for a power aware
+    /// scheduler which toggles between chip power modes based on preexisting knowledge
+    /// of process workloads.
+    unsafe fn execute_kernel_work(&self, chip: &C) {
+        chip.service_pending_interrupts();
+        DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
+    }
+
+    /// Ask the scheduler whether to take a break from executing userspace
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementation, which always prioritizes kernel work, but
+    /// schedulers that wish to defer interrupt handling may reimplement it.
+    unsafe fn break_for_kernel_tasks(&self, kernel: &Kernel, chip: &C) -> bool {
+        chip.has_pending_interrupts()
+            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
+            || kernel.processes_blocked()
+    }
+
+    /// Ask the scheduler whether to return from the loop in `do_process()`. Most
+    /// schedulers will use this default implementation, which causes the do_process()
+    /// loop to return if there are interrupts or deferred calls that need to be serviced.
+    /// However, schedulers which wish to defer interrupt handling may change this, or
+    /// priority schedulers which wish to check if the execution of the current process
+    /// has caused a higher priority process to become ready (such as in the case of IPC)
+    unsafe fn leave_do_process(&self, chip: &C) -> bool {
```

Comment:
```suggestion
    unsafe fn continue_process(&self, process: AppId, chip: &C) -> bool {
```

Name change to try to better reflect what it is asking of the scheduler. Then, do you think it makes sense to tell the kernel what process it is asking about?

Reply:
Agreed, changed.